// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: resources/accounts/oauth2.proto

package accounts

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on OAuth2Account with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *OAuth2Account) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on OAuth2Account with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in OAuth2AccountMultiError, or
// nil if none found.
func (m *OAuth2Account) ValidateAll() error {
	return m.validate(true)
}

func (m *OAuth2Account) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for AccountId

	if utf8.RuneCountInString(m.GetProviderName()) > 255 {
		err := OAuth2AccountValidationError{
			field:  "ProviderName",
			reason: "value length must be at most 255 runes",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if all {
		switch v := interface{}(m.GetProvider()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, OAuth2AccountValidationError{
					field:  "Provider",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, OAuth2AccountValidationError{
					field:  "Provider",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetProvider()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return OAuth2AccountValidationError{
				field:  "Provider",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if utf8.RuneCountInString(m.GetExternalId()) > 128 {
		err := OAuth2AccountValidationError{
			field:  "ExternalId",
			reason: "value length must be at most 128 runes",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if utf8.RuneCountInString(m.GetUsername()) > 255 {
		err := OAuth2AccountValidationError{
			field:  "Username",
			reason: "value length must be at most 255 runes",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if utf8.RuneCountInString(m.GetAvatar()) > 255 {
		err := OAuth2AccountValidationError{
			field:  "Avatar",
			reason: "value length must be at most 255 runes",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if m.CreatedAt != nil {

		if all {
			switch v := interface{}(m.GetCreatedAt()).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, OAuth2AccountValidationError{
						field:  "CreatedAt",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, OAuth2AccountValidationError{
						field:  "CreatedAt",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(m.GetCreatedAt()).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return OAuth2AccountValidationError{
					field:  "CreatedAt",
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return OAuth2AccountMultiError(errors)
	}

	return nil
}

// OAuth2AccountMultiError is an error wrapping multiple validation errors
// returned by OAuth2Account.ValidateAll() if the designated constraints
// aren't met.
type OAuth2AccountMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m OAuth2AccountMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m OAuth2AccountMultiError) AllErrors() []error { return m }

// OAuth2AccountValidationError is the validation error returned by
// OAuth2Account.Validate if the designated constraints aren't met.
type OAuth2AccountValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e OAuth2AccountValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e OAuth2AccountValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e OAuth2AccountValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e OAuth2AccountValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e OAuth2AccountValidationError) ErrorName() string { return "OAuth2AccountValidationError" }

// Error satisfies the builtin error interface
func (e OAuth2AccountValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sOAuth2Account.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = OAuth2AccountValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = OAuth2AccountValidationError{}

// Validate checks the field values on OAuth2Provider with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *OAuth2Provider) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on OAuth2Provider with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in OAuth2ProviderMultiError,
// or nil if none found.
func (m *OAuth2Provider) ValidateAll() error {
	return m.validate(true)
}

func (m *OAuth2Provider) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Name

	// no validation rules for Label

	// no validation rules for Homepage

	if m.Icon != nil {
		// no validation rules for Icon
	}

	if len(errors) > 0 {
		return OAuth2ProviderMultiError(errors)
	}

	return nil
}

// OAuth2ProviderMultiError is an error wrapping multiple validation errors
// returned by OAuth2Provider.ValidateAll() if the designated constraints
// aren't met.
type OAuth2ProviderMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m OAuth2ProviderMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m OAuth2ProviderMultiError) AllErrors() []error { return m }

// OAuth2ProviderValidationError is the validation error returned by
// OAuth2Provider.Validate if the designated constraints aren't met.
type OAuth2ProviderValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e OAuth2ProviderValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e OAuth2ProviderValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e OAuth2ProviderValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e OAuth2ProviderValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e OAuth2ProviderValidationError) ErrorName() string { return "OAuth2ProviderValidationError" }

// Error satisfies the builtin error interface
func (e OAuth2ProviderValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sOAuth2Provider.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = OAuth2ProviderValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = OAuth2ProviderValidationError{}
